### PR TITLE
AIRNEIS-59: Display categories on mobile

### PR DESCRIPTION
**Pull request by cedricgautier**
# Closed PRs : 
- https://github.com/KrispyTech/airneis/pull/62 <- Closed because too many conflicts. 

# DONE

- Fixed get categories and get category by ID endpoint
- Created category page for mobile 
image
image
- Listed categories on homepage so you can test with the category you want
image

# To test 
- Add the env variables from the .env.exemple into your .env file
- Have some categories and products to your DB.
- Launch the project
-  You should be on the home page
-  Go to a category page

### CODE REVIEW

**Review by Magali-D — APPROVED** [review]

File: back/src/controllers/categories/GetCategories.go [hunk 1]
```diff
@@ -30,14 +30,15 @@ func GetCategories(ctx *fiber.Ctx) error {
 func queryCategories(page int, ctx *fiber.Ctx) ([]model.Category, error) {
 	pagLimit := constants.PaginationLimit
 	offset := (page - 1) * pagLimit
+	db := config.Database
 
-	categories := make([]model.Category, pagLimit)
-	errQuery := config.Database.Limit(pagLimit).Offset(offset).Order("order_of_display asc").Find(&categories)
-	if errQuery.Error != nil {
+	cats := make([]model.Category, pagLimit)
+	err := db.Limit(pagLimit).Preload("Products").Offset(offset).Order("order_of_display asc").Find(&cats)
```
Comment (cedricgautier) [inline back/src/controllers/categories/GetCategories.go]: I think it would be useful to refactor this function and make it available for multiple of our cases.

File: back/src/controllers/categories/GetCategories.go [hunk 2]
```diff
@@ -30,14 +30,15 @@ func GetCategories(ctx *fiber.Ctx) error {
 func queryCategories(page int, ctx *fiber.Ctx) ([]model.Category, error) {
 	pagLimit := constants.PaginationLimit
 	offset := (page - 1) * pagLimit
+	db := config.Database
 
-	categories := make([]model.Category, pagLimit)
-	errQuery := config.Database.Limit(pagLimit).Offset(offset).Order("order_of_display asc").Find(&categories)
-	if errQuery.Error != nil {
+	cats := make([]model.Category, pagLimit)
```
Comment (cedricgautier) [inline back/src/controllers/categories/GetCategories.go]: Cats are cool, but you make categories right ? 
image

File: back/src/controllers/categories/categories.go [hunk 3]
```diff
@@ -38,5 +39,5 @@ func sendCategories(ctx *fiber.Ctx, categories []model.Category) error {
 func SelectCategoryByID(category *model.Category, ctx *fiber.Ctx) error {
 	categoryID := ctx.Params("categoryID")
 
-	return config.Database.First(&category, categoryID).Error
+	return config.Database.Preload("Products").First(&category, categoryID).Error
```
Comment (cedricgautier) [inline back/src/controllers/categories/categories.go:42]: This also could be used to make a global helper to SelectModelByID, what do you think ?

File: mobile/src/components/Header.jsx [hunk 4]
```diff
@@ -7,9 +7,7 @@ import screens from "../utils/constants/screens"
 
 const Header = props => {
   const { navigation } = props
-  const goToHomePage = () => {
-    navigation.navigate(screens.homePage)
-  }
+  const goToHomePage = () => navigation.navigate(screens.homePage)
```
Comment (cedricgautier) [inline mobile/src/components/Header.jsx]: Can you please name this : `navigateToHomePage` I think it would be clearer

File: back/src/models/Category.go [hunk 5]
```diff
@@ -4,8 +4,10 @@ import "gorm.io/gorm"
 
 type Category struct {
 	gorm.Model
-	Name           string `json:"name"`
-	ThumbnailURL   string `json:"thumbnailURL"`
-	Slug           string `json:"slug"           gorm:"unique"`
-	OrderOfDisplay *uint  `json:"orderOfDisplay" gorm:"unique"`
+	Name           string    `json:"name"`
+	ThumbnailURL   string    `json:"thumbnailUrl"`
```
Comment (cedricgautier) [inline back/src/models/Category.go]: Why did you modify this ? Keep it as it was

File: mobile/mobileConfig.js [hunk 6]
```diff
@@ -1,4 +1,17 @@
 const mobileConfig = {
+  api: {
+    url: process.env.EXPO_PUBLIC_API_URL
+  },
+  images: {
+    fullScreenImageXl: {
+      width: "100%",
+      height: 300
+    },
+    fullScreenImageMd: {
```
Comment (cedricgautier) [inline mobile/mobileConfig.js]: it's MD not `Md`, same for `XL`

File: back/src/controllers/categories/GetCategories.go [hunk 7]
```diff
@@ -30,14 +30,15 @@ func GetCategories(ctx *fiber.Ctx) error {
...
+	err := db.Limit(pagLimit).Preload("Products").Offset(offset).Order("order_of_display asc").Find(&cats)
+	if err.Error != nil {
 		ctx.Status(constants.BadRequestStatus)
 
-		return categories, ctx.SendString(constants.BadRequestMessage)
+		return cats, ctx.SendString(constants.BadRequestMessage)
 	}
 
-	return categories, nil
+	return cats, nil
```
Comment (cedricgautier) [inline back/src/controllers/categories/GetCategories.go]: You're returning Cats here, :trollface:

File: back/src/controllers/categories/categories.go (hunk 3 above)
Comment (cedricgautier) [reply back/src/controllers/categories/categories.go:42]: You can make a helper just for database actions

File: back/src/models/Category.go (hunk 5 above)
Comment (cedricgautier) [reply back/src/models/Category.go]: You have created an inconsistency here : 
- https://github.com/KrispyTech/airneis/pull/78/files#diff-328223e6b0390f7ee84d18be3eccdb3e59c201aa158ca9f54d8ca2f77465edccR46
- https://github.com/KrispyTech/airneis/pull/78/files#diff-ba1d263262df9422533f8a0063995b401182fcfde22a9fe563e16243d28fc249R6

File: mobile/src/screens/HomeScreen.jsx [hunk 8]
```diff
@@ -1,15 +1,40 @@
...
+import { getCategories } from "../utils/categoriesServices"
+import { useEffect, useState } from "react"
+import ErrorScreen from "./ErrorScreen"
+import errors from "../utils/constants/errors"
 
 const HomeScreen = props => {
   const { navigation } = props
-  const goToCategory = () => navigation.navigate(screens.categoryPage)
+  const [categories, setCategories] = useState([])
+  const [hadError, setHadError] = useState(false)
```
Comment (cedricgautier) [inline mobile/src/screens/HomeScreen.jsx]: I think it would be better to have something like : ` const [hasError, setError] = useState(false)`

- You're setting an error and something had an error

File: mobile/src/screens/HomeScreen.jsx [hunk 9]
```diff
@@ -1,15 +1,40 @@
...
+  const goToCategory = id => () => navigation.navigate(screens.categoryPage, { id })
+
+  useEffect(() => {
+    ;(async () => {
+      try {
+        const categoriesResult = await getCategories()
+        setCategories(categoriesResult.data)
+      } catch (e) {
```
Comment (cedricgautier) [inline mobile/src/screens/HomeScreen.jsx:20]: You catch e here, where is it used ?

File: mobile/src/screens/HomeScreen.jsx [hunk 10]
```diff
@@ -1,15 +1,40 @@
...
+        setHadError(true)
+      }
+    })()
+  }, [])
+
+  if (hadError) {
```
Comment (cedricgautier) [inline mobile/src/screens/HomeScreen.jsx:26]: change to `hasError`

File: back/src/controllers/categories/GetCategories.go (hunk 1 above)
Comment (barnabepilliaudin) [reply back/src/controllers/categories/GetCategories.go]: Good idea I will try to implement it

File: mobile/src/screens/HomeScreen.jsx (hunk 9 above)
Comment (barnabepilliaudin) [reply mobile/src/screens/HomeScreen.jsx:20]: It throw an error if the brackets are empty

File: back/src/controllers/categories/GetCategories.go (hunk 1 above)
Comment (barnabepilliaudin) [reply back/src/controllers/categories/GetCategories.go]: Let's do it in an other PR later so it doesn't block the mobile

**Review by Magali-D — APPROVED** [review]
I like it, I think it's clearer ! Gj!

File: mobile/src/screens/HomeScreen.jsx (hunk 9 above)
Comment (cedricgautier) [reply mobile/src/screens/HomeScreen.jsx:20]: Yes but you are not catching "e", why ?

Comment (barnabepilliaudin) [reply mobile/src/screens/HomeScreen.jsx:20]: You must catch e if you don't want an error to be thrown. I tried it 
Mb in previous comment I said braquets instead of parentheses

File: mobile/src/screens/CategoryScreen.jsx [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  return (
+    <Page navigation={navigation}>
+      <ScrollView className="relative">
+        <View>
+          <Image
+            style={mobileConfig.images.fullScreenImageXL}
+            source={{
+              uri: category.thumbnailURL
```
Comment (cedricgautier) [inline mobile/src/screens/CategoryScreen.jsx]: You've written `uri` instead of `URL`

File: mobile/src/screens/HomeScreen.jsx (hunk 9 above)
Comment (cedricgautier) [reply mobile/src/screens/HomeScreen.jsx:20]: But the variable "e", aren't you going to use it ?

File: back/src/controllers/categories/GetCategories.go [hunk 12]
```diff
@@ -30,10 +30,11 @@ func GetCategories(ctx *fiber.Ctx) error {
 func queryCategories(page int, ctx *fiber.Ctx) ([]model.Category, error) {
 	pagLimit := constants.PaginationLimit
 	offset := (page - 1) * pagLimit
+	db := config.Database
 
 	categories := make([]model.Category, pagLimit)
-	errQuery := config.Database.Limit(pagLimit).Offset(offset).Order("order_of_display asc").Find(&categories)
-	if errQuery.Error != nil {
+	err := db.Limit(pagLimit).Preload("Products").Offset(offset).Order("order_of_display asc").Find(&categories)
```
Comment (cedricgautier) [inline back/src/controllers/categories/GetCategories.go:36]: Why are you preloading "Products", aren't you returning categories ?

File: back/src/controllers/categories/categories.go (hunk 3 above)
Comment (cedricgautier) [reply back/src/controllers/categories/categories.go:42]: As discussed, this could be later on refactored in another MR

File: back/src/controllers/categories/GetCategories.go (hunk 12 above)
Comment (barnabepilliaudin) [reply back/src/controllers/categories/GetCategories.go:36]: Preloading products allows to have the products with all the details in the response. It's like with graph fetch ok Objection

File: mobile/src/screens/CategoryScreen.jsx (hunk 11 above)
Comment (barnabepilliaudin) [reply mobile/src/screens/CategoryScreen.jsx]: Done thx. I thought it won't work but it did

File: mobile/src/screens/HomeScreen.jsx (hunk 9 above)
Comment (barnabepilliaudin) [reply mobile/src/screens/HomeScreen.jsx:20]: No has the only thing that can go wrong on this screen is the loading of the categories, if an error is thrown it means  that the category loading failed so no need to use it for the moment. But still need to to catch it to avoid an error. 

But this can be the subject of a rework

File: back/src/controllers/categories/GetCategories.go (hunk 12 above)
Comment (cedricgautier) [reply back/src/controllers/categories/GetCategories.go:36]: Sounds good, thanks

**Comment by cedricgautier** [comment]
Looks good to me, I can't approve because I created the MR to resolve issues on MR creation even though @BarnabePILLIAUDIN is the author of the MR. 

TLDR : I approve, this can be merged

**Review by BarnabePILLIAUDIN — APPROVED** [review]
